File: rn/modules/canvas-api/apis/courses.js
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

/* @flow */

import { paginate, exhaust } from '../utils/pagination'
import httpClient from '../httpClient'

export function getCourses (): Promise<ApiResponse<Course[]>> {
  const courses = paginate('courses', {
    params: {
      include: ['term', 'favorites', 'course_image', 'sections'],
    },
  })
  return exhaust(courses)
}

export function getCourseTabs (courseID: string): Promise<ApiResponse<Tab[]>> {
  const tabs = paginate(`courses/${courseID}/tabs`, {
    per_page: 99,
  })
  return exhaust(tabs)
}

export function updateCourse (course: Course): Promise<ApiResponse<Course>> {
  const safeParams = {
    name: course.name,
    default_view: course.default_view,
  }
  return httpClient().put(`/courses/${course.id}`, { course: safeParams })
}

export function getCourseAssignments (courseID: string): Promise<ApiResponse<Assignment[]>> {
  const url = `courses/${courseID}/assignments`
  return httpClient().get(url)
}

export function getCourseGradingPeriods (courseID: string): Promise<ApiResponse<GradingPeriodResponse>> {
  const url = `courses/${courseID}/grading_periods`
  return httpClient().get(url)
}

export function getCourseSections (courseID: string): Promise<ApiResponse<Section[]>> {
  const url = `courses/${courseID}/sections`
  const options = {
    params: {
      include: ['total_students'],
    },
  }
  return httpClient().get(url, options)
}

export function getCourseUsers (courseID: string, enrollmentType?: string = ''): Promise<ApiResponse<User[]>> {
  const url = `courses/${courseID}/users`
  const params = {}
  if (enrollmentType) {
    params.enrollment_type = [enrollmentType]
  }
  const options = { params }
  return httpClient().get(url, options)
}

export function favoriteCourse (courseID: string): Promise<AxiosResponse<Favorite>> {
  return httpClient().post(`users/self/favorites/courses/${courseID}`)
}

export function unfavoriteCourse (courseID: string): Promise<AxiosResponse<Favorite>> {
  return httpClient().delete(`users/self/favorites/courses/${courseID}`)
}

export function updateCourseColor (courseId: string, hexcode: string): Promise<AxiosResponse<UpdateCustomColorResponse>> {
  return httpClient().put(`users/self/colors/course_${courseId}`, { hexcode })
}

File: Canvas/GradesWidget/Network/Courses.swift
//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

import Foundation
import CanvasKeymaster

let CoursesCacheKey = "GradesWidgetCacheCoursesKey"

func getCourses(client: CKIClient, completionHandler: @escaping (Result<[Course]>) -> Void) {
    if let cache = loadCache() {
        completionHandler(.completed(cache))
    }

    guard let baseURL = client.baseURL, let token = client.accessToken else {
        completionHandler(.failed(NetworkError.invalidRequest))
        return
    }
    let url = baseURL.appendingPathComponent("api/v1/courses")
    var urlComponents = URLComponents(url: url, resolvingAgainstBaseURL: false)
    urlComponents?.queryItems = [
        URLQueryItem(name: "include[]", value: "favorites"),
        URLQueryItem(name: "include[]", value: "total_scores"),
        URLQueryItem(name: "include[]", value: "current_grading_period_scores"),
        URLQueryItem(name: "per_page", value: "99") 
    ]
    var request = URLRequest(url: urlComponents?.url ?? url)
    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    request.setValue("application/json+canvas-string-ids", forHTTPHeaderField: "Accept")

    let task = URLSession.shared.dataTask(with: request) { (data, response, error) in
        guard let httpResponse = response as? HTTPURLResponse, (200..<300).contains(httpResponse.statusCode), let data = data else {
            completionHandler(.failed(error ?? NetworkError.invalidResponse))
            return
        }

        do {
            let decoder = JSONDecoder()
            let courses = try decoder.decode([Course].self, from: data)
            let favorites = courses.filter { $0.isFavorite }
            try writeToCache(favorites)
            completionHandler(.completed(favorites))
            return
        } catch let e {
            completionHandler(.failed(e))
            return
        }
    }

    task.resume()
}

private func loadCache() -> [Course]? {
    guard let coursesData = UserDefaults.standard.data(forKey: CoursesCacheKey) else {
        return nil
    }
    let decoder = JSONDecoder()
    return try? decoder.decode([Course].self, from: coursesData)
}

private func writeToCache(_ courses: [Course]) throws {
    let encoder = JSONEncoder()
    let encoded = try encoder.encode(courses)
    UserDefaults.standard.set(encoded, forKey: CoursesCacheKey)
}
